analyzer/protocol/stun/stun.spicy

```diff
@@ -0,0 +1,215 @@
+#                      |11|10|9|8|7|1|6|5|4|0|3|2|1|0|
+#                      +--+--+-+-+-+-+-+-+-+-+-+-+-+-+
+
+public type STUNPacket = unit {
+	var M: uint16;
+	var C: uint16;
+
+	msgtype: bitfield(16) {
+	  m0: 0..3;
+	  c0: 4;
+	  m1: 5..7;
+	  c1: 8;
+	  m2: 9..13;
+	  zeros: 14..15 &requires=($$ == 0);
+	};
+
+	on msgtype {
+		self.M = self.msgtype.m0 + 16*self.msgtype.m1 + 128*self.msgtype.m2;
+		self.C = self.msgtype.c0 + 2*self.msgtype.c1;
+	}
+
+	msg_length: uint16;
+	cookie: uint32 &requires=($$ == 554869826);
+	trans_id: bytes &size=12;
+	attributes: Attribute(self.M, self.C, self.trans_id)[] &size=self.msg_length;
+};
+
+#       0                   1                   2                   3
+#       0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#      |         Type                  |            Length             |
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#      |                         Value (variable)...                   |
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#      https://tools.ietf.org/html/rfc5389#section-15
+
+type Attribute = unit(method: uint16, class: uint16, trans_id: bytes) {
+	var trans_id: bytes = trans_id;
+	var method: uint16 = method;
+	var class: uint16 = class;
+	attr_type: uint16 &convert=AttrType($$);
+	attr_length: uint16 {
+		self.value_padding = padding($$);
+		self.total_size = $$ + self.value_padding;
+	}
+	switch (self.attr_type) {
+		AttrType::MAPPED_ADDRESS -> MAPPED_ADDRESS: MappedAddress(self.attr_length);
+		AttrType::USERNAME -> USERNAME: bytes &size=self.attr_length;
+		AttrType::MESSAGE_INTEGRITY -> HMAC_SHA1: bytes &size=20;
+		AttrType::ERROR_CODE -> ERROR_CODE: ErrorCode(self.attr_length);
+		AttrType::UNKNOWN_ATTRIBUTES -> UNKNOWN_ATTRIBUTES: UnknownAttributes(self.attr_length);
+		AttrType::REALM -> REALM: bytes &size=self.attr_length;
+		AttrType::NONCE -> NONCE: bytes &size=self.attr_length;
+		AttrType::XOR_MAPPED_ADDRESS -> XOR_MAPPED_ADDRESS: MappedAddress(self.attr_length);
+		AttrType::XOR_PEER_ADDRESS -> XOR_PEER_ADDRESS: MappedAddress(self.attr_length);
+		AttrType::XOR_RELAYED_ADDRESS -> XOR_RELAYED_ADDRESS: MappedAddress(self.attr_length);
+		AttrType::PRIORITY -> PRIORITY: bytes &size=4;
+		AttrType::USE_CANDIDATE -> USE_CANDIDATE: bytes &size=0;
+		AttrType::SOFTWARE -> DESCRIPTION: bytes &size=self.attr_length;
+		AttrType::ALTERNATE_SERVER -> ALTERNATE_SERVER: MappedAddress(self.attr_length);
+		AttrType::FINGERPRINT -> CRC_32: bytes &size=self.attr_length;
+		AttrType::ICE_CONTROLLED -> ICE_CONTROLLED: bytes &size=self.attr_length;
+		AttrType::ICE_CONTROLLING -> ICE_CONTROLLING: bytes &size=self.attr_length;
+		* -> attr_value: bytes &size=self.total_size;
+	};
+	padding: bytes &size=self.value_padding;
+
+	var value_padding: uint16;
+	var total_size: uint16;
```

Comment:
Let's move these up to the other `var` decls so it isn't hidden down here.

Reply:
moved